File: javascipt/caesar_cipher.js
// Write a function that takes a message and an increment amount and
// outputs the same letters shifted by that amount in the alphabet. 
// Assume lowercase and no punctuation. Preserve spaces.

function caesarCipher(str, shift) {
  return str.split(" ").map(word => caesarWord(word, shift)).join(" ");
};

function caesarWord(word, shift) {
  let letters = 'abcdefghijklmnopqrstuvwxyz'
  let uLetters = letters.toUpperCase();
  letters += uLetters;

  let oldIdx, newIdx;
  let newWord =  word.split("").map(ch => {
    oldIdx = letters.indexOf(ch);
    newIdx = (oldIdx % 26 + shift) % 26;
    newIdx += oldIdx < 26 ? 0 : 26;

    return letters[newIdx];
  }).join("");

  return newWord;
};

console.log(caesarCipher('abc xyz ABC XYZ', 1));
// 'bcd yza BCD YZA'

File: ruby/caesar_cipher.rb
# Write a function that takes a message and an increment amount and
# outputs the same letters shifted by that amount in the alphabet. 
# Assume lowercase and no punctuation. Preserve spaces.

def caesar_cipher(str, shift)
  str.split.map { |word| caesar_cipher_word(word, shift) }.join(" ")
end

def caesar_cipher_word(word, shift)
  letters = ('a'..'z').to_a + ('A'..'Z').to_a
  word.chars.map do |ch|
    old_idx = letters.index(ch)
    new_idx = (old_idx % 26 + shift) % 26
    new_idx += old_idx < 26 ? 0 : 26

    letters[new_idx]
  end.join
end


a = "abcxxxABC"
p caesar_cipher(a, 3) # 'defaaaDEF'